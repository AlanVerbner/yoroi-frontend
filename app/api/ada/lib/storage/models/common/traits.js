// @flow


import type {
  lf$Transaction,
} from 'lovefield';

import {
  Mixin,
} from 'mixwith';

import type {
  IPublicDeriver,
  IGetPublic, IGetPublicRequest, IGetPublicResponse,
  IGetAllUtxos,
  IGetUtxoBalance, IGetUtxoBalanceRequest, IGetUtxoBalanceResponse,
  IScanAddresses,
} from '../PublicDeriver/interfaces';
import type {
  IGetBalance, IGetBalanceRequest, IGetBalanceResponse,
  IChangePasswordRequest, IChangePasswordResponse,
} from './interfaces';
import { ConceptualWallet } from '../ConceptualWallet/index';
import type { IConceptualWallet } from '../ConceptualWallet/interfaces';
import type {
  IHasLevels, IHasSign, IHasPrivateDeriver,
} from './wrapper/interfaces';

import {
  getBalanceForUtxos,
  rawChangePassword,
} from '../utils';

import {
  getAllSchemaTables,
  raii,
  StaleStateError,
  mapToTables,
} from '../../database/utils';

import {
  GetKeyForPublicDeriver,
} from '../../database/walletTypes/core/api/read';
import { GetDerivationSpecific } from '../../database/walletTypes/common/api/read';

import {
  GetUtxoTxOutputsWithTx,
} from  '../../database/transactionModels/utxo/api/read';

import {
  GetPathWithSpecific,
  GetAddress,
} from '../../database/primitives/api/read';
import { UpdateGet, } from '../../database/primitives/api/write';

// =================
//   GetPublicKey
// =================

type GetPublicKeyDependencies = IPublicDeriver<>;
const GetPublicKeyMixin = (
  superclass: Class<GetPublicKeyDependencies>,
) => class GetPublicKey extends superclass implements IGetPublic {

  rawGetPublicKey: (
    lf$Transaction,
    {| GetKeyForPublicDeriver: Class<GetKeyForPublicDeriver> |},
    IGetPublicRequest,
  ) => Promise<IGetPublicResponse> = async (tx, deps, _body) => {
    const derivationAndKey = await deps.GetKeyForPublicDeriver.get(
      super.getDb(), tx,
      super.getPublicDeriverId(),
      true,
      false,
    );
    if (derivationAndKey.publicKey == null) {
      throw new StaleStateError('GetPublicKey::rawGetPublicKey publicKey');
    }
    return derivationAndKey.publicKey;
  }
  getPublicKey: IGetPublicRequest => Promise<IGetPublicResponse> = async (body) => {
    const deps = Object.freeze({
      GetKeyForPublicDeriver,
    });
    const depTables = Object
      .keys(deps)
      .map(key => deps[key])
      .flatMap(table => getAllSchemaTables(super.getDb(), table));
    return await raii(
      super.getDb(),
      depTables,
      async tx => this.rawGetPublicKey(tx, deps, body)
    );
  }

  rawChangePubDeriverPassword: (
    lf$Transaction,
    {|
      UpdateGet: Class<UpdateGet>,
      GetKeyForPublicDeriver: Class<GetKeyForPublicDeriver>
    |},
    IChangePasswordRequest,
  ) => Promise<IChangePasswordResponse> = async (tx, deps, body) => {
    const currentRow = await this.rawGetPublicKey(
      tx,
      { GetKeyForPublicDeriver: deps.GetKeyForPublicDeriver, },
      undefined,
    );
    return rawChangePassword(
      super.getDb(), tx,
      { UpdateGet: deps.UpdateGet, },
      {
        ...body,
        oldKeyRow: currentRow
      },
    );
  }
  changePubDeriverPassword: IChangePasswordRequest => Promise<IChangePasswordResponse> = async (body) => {
    const deps = Object.freeze({
      UpdateGet,
      GetKeyForPublicDeriver,
    });
    const depTables = Object
      .keys(deps)
      .map(key => deps[key])
      .flatMap(table => getAllSchemaTables(super.getDb(), table));
    return await raii(
      super.getDb(),
      depTables,
      async tx => this.rawChangePubDeriverPassword(tx, deps, body)
    );
  }
};
export const GetPublicKey = Mixin<
  GetPublicKeyDependencies,
  IGetPublic
>(GetPublicKeyMixin);
const GetPublicKeyInstance = (
  (GetPublicKey: any): ReturnType<typeof GetPublicKeyMixin>
);
export function asGetPublicKey<T: IPublicDeriver<any>>(
  obj: T
): void | (IGetPublic & GetPublicKeyDependencies & T) {
  if (obj instanceof GetPublicKeyInstance) {
    return obj;
  }
  return undefined;
}

// ==============
//   GetBalance
// ==============

type GetBalanceDependencies = IPublicDeriver<> & IGetUtxoBalance;
const GetBalanceMixin = (
  superclass: Class<GetBalanceDependencies>,
) => class GetBalance extends superclass implements IGetBalance {
  getBalance: IGetBalanceRequest => Promise<IGetBalanceResponse> = async (body) => {
    return await this.getUtxoBalance(body);
  }
};

export const GetBalance = Mixin<
  GetBalanceDependencies,
  IGetBalance,
>(GetBalanceMixin);
const GetBalanceInstance = (
  (GetBalance: any): ReturnType<typeof GetBalanceMixin>
);
export function asGetBalance<T: IPublicDeriver<any>>(
  obj: T
): void | (IGetBalance & GetBalanceDependencies & T) {
  if (obj instanceof GetBalanceInstance) {
    return obj;
  }
  return undefined;
}


// ==================
//   GetUtxoBalance
// ==================

type GetUtxoBalanceDependencies = IPublicDeriver<ConceptualWallet & IHasLevels> & IGetAllUtxos;
const GetUtxoBalanceMixin = (
  superclass: Class<GetUtxoBalanceDependencies>,
) => class GetUtxoBalance extends superclass implements IGetUtxoBalance {

  rawGetUtxoBalance: (
    lf$Transaction,
    {|
      GetPathWithSpecific: Class<GetPathWithSpecific>,
      GetAddress: Class<GetAddress>,
      GetUtxoTxOutputsWithTx: Class<GetUtxoTxOutputsWithTx>,
      GetDerivationSpecific: Class<GetDerivationSpecific>,
    |},
    IGetUtxoBalanceRequest,
    Map<number, string>,
  ) => Promise<IGetUtxoBalanceResponse> = async (tx, deps, _body, derivationTables) => {
    const utxos = await this.rawGetAllUtxos(
      tx,
      {
        GetAddress: deps.GetAddress,
        GetPathWithSpecific: deps.GetPathWithSpecific,
        GetUtxoTxOutputsWithTx: deps.GetUtxoTxOutputsWithTx,
        GetDerivationSpecific: deps.GetDerivationSpecific,
      },
      undefined,
      derivationTables,
    );
    return getBalanceForUtxos(utxos.map(utxo => utxo.output.UtxoTransactionOutput));
  }
  getUtxoBalance: IGetUtxoBalanceRequest => Promise<IGetUtxoBalanceResponse> = async (_body) => {
    const derivationTables = this.getParent().getDerivationTables();
    const deps = Object.freeze({
      GetPathWithSpecific,
      GetAddress,
      GetUtxoTxOutputsWithTx,
      GetDerivationSpecific,
    });
    const depTables = Object
      .keys(deps)
      .map(key => deps[key])
      .flatMap(table => getAllSchemaTables(super.getDb(), table));
    return await raii<IGetUtxoBalanceResponse>(
      super.getDb(),
      [
        ...depTables,
        ...mapToTables(super.getDb(), derivationTables),
      ],
      async tx => this.rawGetUtxoBalance(tx, deps, undefined, derivationTables)
    );
  }
};

export const GetUtxoBalance = Mixin<
  GetUtxoBalanceDependencies,
  IGetUtxoBalance,
>(GetUtxoBalanceMixin);
const GetUtxoBalanceInstance = (
  (GetUtxoBalance: any): ReturnType<typeof GetUtxoBalanceMixin>
);
export function asGetUtxoBalance<T: IPublicDeriver<any>>(
  obj: T
): void | (IGetUtxoBalance & GetUtxoBalanceDependencies & T) {
  if (obj instanceof GetUtxoBalanceInstance) {
    return obj;
  }
  return undefined;
}

// =================
//   ScanAddresses
// =================

type ScanAddressesDependencies = IPublicDeriver<> & IScanAddresses;
const ScanAddressesMixin = (
  superclass: Class<ScanAddressesDependencies>,
) => class ScanAddresses extends superclass implements IScanAddresses {
};

export const ScanAddresses = Mixin<
  ScanAddressesDependencies,
  IScanAddresses,
>(ScanAddressesMixin);
const ScanAddressesInstance = (
  (ScanAddresses: any): ReturnType<typeof ScanAddressesMixin>
);
export function asScanAddresses<T: IPublicDeriver<any>>(
  obj: T
): void | (IScanAddresses & ScanAddressesDependencies & T) {
  if (obj instanceof ScanAddressesInstance) {
    return obj;
  }
  return undefined;
}

type HasPrivateDeriverDependencies = IPublicDeriver<ConceptualWallet & IHasPrivateDeriver>;
const HasPrivateDeriverMixin = (
  superclass: Class<HasPrivateDeriverDependencies>,
) => class HasPrivateDeriver extends superclass {
};
export const HasPrivateDeriver = Mixin<
  HasPrivateDeriverDependencies,
  Object,
>(HasPrivateDeriverMixin);
export function asHasPrivateDeriver<Wrapper: ConceptualWallet, Rest>(
  obj: IPublicDeriver<Wrapper> & Rest
): void | (IPublicDeriver<Wrapper & IHasLevels> & Rest) {
  if (obj instanceof HasPrivateDeriver) {
    return obj;
  }
  return undefined;
}

type HasLevelsDependencies = IPublicDeriver<ConceptualWallet & IHasLevels>;
const HasLevelsMixin = (
  superclass: Class<HasLevelsDependencies>,
) => class HasLevels extends superclass {
};
export const HasLevels = Mixin<
  HasLevelsDependencies,
  Object,
>(HasLevelsMixin);
export function asHasLevels<Wrapper: ConceptualWallet, Rest>(
  obj: IPublicDeriver<Wrapper> & Rest
): void | (IPublicDeriver<Wrapper & IHasLevels> & Rest) {
  if (obj instanceof HasLevels) {
    return obj;
  }
  return undefined;
}

type HasSignDependencies = IPublicDeriver<ConceptualWallet & IHasSign>;
const HasSignMixin = (
  superclass: Class<HasSignDependencies>,
) => class HasSign extends superclass {
};
export const HasSign = Mixin<
  HasSignDependencies,
  Object,
>(HasSignMixin);
export function asHasSign<Wrapper: ConceptualWallet, Rest>(
  obj: IPublicDeriver<Wrapper> & Rest
): void | (IPublicDeriver<Wrapper & IHasSign> & Rest) {
  if (obj instanceof HasSign) {
    return obj;
  }
  return undefined;
}
