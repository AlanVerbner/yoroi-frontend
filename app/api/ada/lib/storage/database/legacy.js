// @flow

import { dumpByVersion } from './index';
import { RustModule } from '../../cardanoCrypto/rustLoader';

/**
 * This file contains methods used to extract information
 * from the legacy database format
 * They should NOT be used for any purpose other than
 * to migrate to a new format
 */

export type LegacyAddressingInfo = {
  account: number,
  change: number,
  index: number,
};
export type LegacyAdaAmount = {
  getCCoin: string,
};
export type LegacyAdaAddress = {
  cadAmount: LegacyAdaAmount,
  cadId: string,
  cadIsUsed: boolean,
} & LegacyAddressingInfo;

type LegacyLocalStorageWallet = {
  adaWallet: AdaWallet,
  masterKey?: string, // unused in hardware wallets
  // this is a per-account setting but we only have 1 account per wallet in Yoroi
  lastReceiveAddressIndex: number
}

export type LegacyCryptoAccount = {
  account: number,
  // master public key
  root_cached_key: RustModule.Wallet.Bip44AccountPublic,
  derivation_scheme: string
}

type LegacyLocalStorageCryptoAccount = {
  account: number,
  root_cached_key: string, // MasterPublicKey
  derivation_scheme: string
}

export const getLegacyAddressesList = (): Array<LegacyAdaAddress> => {
  if (dumpByVersion.Addresses) {
    return dumpByVersion.Addresses;
  }
  return [];
};

export const resetLegacy = (): void => {
  for (const prop of Object.keys(dumpByVersion)) {
    delete dumpByVersion[prop];
  }
};
