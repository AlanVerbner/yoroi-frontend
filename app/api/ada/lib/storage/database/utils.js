// @flow

import type {
  lf$Database,
  lf$Row,
  lf$Transaction,
} from 'lovefield';

export async function addToTable<Insert, Row>(
  db: lf$Database,
  tx: lf$Transaction,
  request: Insert,
  tableName: string,
): Promise<Row> {
  const table = db.getSchema().table(tableName);
  const newRow = table.createRow(request);

  const result: Row = (await tx.attach(
    db
      .insertOrReplace()
      .into(table)
      .values(([newRow]: Array<lf$Row>))
  ))[0];

  return result;
}

export const getRowFromKey = async <T>(
  db: lf$Database,
  tx: lf$Transaction,
  key: number,
  tableName: string,
  keyRowName: string,
): Promise<T | void> => {
  const table = db.getSchema().table(tableName);
  const query = db
    .select()
    .from(table)
    .where(table[keyRowName].eq(key));
  const result = await tx.attach(query);
  if (result.length === 0) {
    return undefined;
  }
  return result[0];
};

export async function getRowIn<Row>(
  db: lf$Database,
  tx: lf$Transaction, 
  tableName: string,
  keyRowName: string,
  list: Array<any>,
): Promise<Array<Row>> {
  const table = db.getSchema().table(tableName);
  const query = db
    .select()
    .from(table)
    .where(table[keyRowName].in(list));
  return await tx.attach(query);
}
