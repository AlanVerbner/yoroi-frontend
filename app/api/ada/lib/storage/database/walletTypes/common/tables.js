// @flow

import {
  AddressSchema,
  KeyDerivationSchema,
} from '../../primitives/tables';
import { Bip44WrapperSchema } from '../bip44/tables';
import { Type } from 'lovefield';
import type { lf$schema$Builder } from 'lovefield';

export type PrivateDeriverInsert = {|
  Bip44WrapperId: number,
  KeyDerivationId: number,
  Level: number,
|};
export type PrivateDeriverRow = {|
  PrivateDeriverId: number, // serial
  ...PrivateDeriverInsert,
|};
export const PrivateDeriverSchema: {
  +name: 'PrivateDeriver',
  properties: $ObjMapi<PrivateDeriverRow, ToSchemaProp>
} = {
  name: 'PrivateDeriver',
  properties: {
    PrivateDeriverId: 'PrivateDeriverId',
    Bip44WrapperId: 'Bip44WrapperId',
    KeyDerivationId: 'KeyDerivationId',
    Level: 'Level',
  }
};

export type Bip44RootInsert = {|
  KeyDerivationId: number,
|};
export type Bip44RootRow = {|
  Bip44RootId: number,
  ...Bip44RootInsert,
|};
export const Bip44RootSchema: {
  +name: 'Bip44Root',
  properties: $ObjMapi<Bip44RootRow, ToSchemaProp>
} = {
  name: 'Bip44Root',
  properties: {
    Bip44RootId: 'Bip44RootId',
    KeyDerivationId: 'KeyDerivationId',
  }
};
export type Bip44PurposeInsert = {|
  KeyDerivationId: number,
|};
export type Bip44PurposeRow = {|
  Bip44PurposeId: number,
  ...Bip44PurposeInsert,
|};
export const Bip44PurposeSchema: {
  +name: 'Bip44Purpose',
  properties: $ObjMapi<Bip44PurposeRow, ToSchemaProp>
} = {
  name: 'Bip44Purpose',
  properties: {
    Bip44PurposeId: 'Bip44PurposeId',
    KeyDerivationId: 'KeyDerivationId',
  }
};
export type Bip44CoinTypeInsert = {|
  KeyDerivationId: number,
|};
export type Bip44CoinTypeRow = {|
  Bip44CoinTypeId: number,
  ...Bip44CoinTypeInsert,
|};
export const Bip44CoinTypeSchema: {
  +name:'Bip44CoinType',
  properties: $ObjMapi<Bip44CoinTypeRow, ToSchemaProp>
} = {
  name: 'Bip44CoinType',
  properties: {
    Bip44CoinTypeId: 'Bip44CoinTypeId',
    KeyDerivationId: 'KeyDerivationId',
  }
};
export type Bip44AccountInsert = {|
  KeyDerivationId: number,
|};
export type Bip44AccountRow = {|
  Bip44AccountId: number,
  ...Bip44AccountInsert,
|};
export const Bip44AccountSchema: {
  +name: 'Bip44Account',
  properties: $ObjMapi<Bip44AccountRow, ToSchemaProp>
} = {
  name: 'Bip44Account',
  properties: {
    Bip44AccountId: 'Bip44AccountId',
    KeyDerivationId: 'KeyDerivationId',
  }
};
export type Bip44ChainInsert = {|
  KeyDerivationId: number,
  DisplayCutoff: number | null,
|};
export type Bip44ChainRow = {|
  Bip44ChainId: number,
  ...Bip44ChainInsert,
|};
export const Bip44ChainSchema: {
  +name: 'Bip44Chain',
  properties: $ObjMapi<Bip44ChainRow, ToSchemaProp>
} = {
  name: 'Bip44Chain',
  properties: {
    Bip44ChainId: 'Bip44ChainId',
    KeyDerivationId: 'KeyDerivationId',
    DisplayCutoff: 'DisplayCutoff',
  }
};
export type Bip44AddressInsert = {|
  KeyDerivationId: number,
  AddressId: number,
|};
export type Bip44AddressRow = {|
  Bip44AddressId: number,
  ...Bip44AddressInsert,
|};
export const Bip44AddressSchema: {
  +name: 'Bip44Address',
  properties: $ObjMapi<Bip44AddressRow, ToSchemaProp>
} = {
  name: 'Bip44Address',
  properties: {
    Bip44AddressId: 'Bip44AddressId',
    KeyDerivationId: 'KeyDerivationId',
    /**
     * We need to specify an index into another table instead of storing the hash here directly
     * This is because we need an address table entry for every input & output in a transaction
     * even if it doesn't belong to you.
     * We can't make that a foreign key to this table because this table has a "KeyDerivationId"
     * We can't make the "KeyDerivationId" nullable because you can't create an index on a nullable
     */
    AddressId: 'AddressId',
  }
};

export type AccountingDerivationInsert = {|
  KeyDerivationId: number,
  AddressId: number,
|};
export type AccountingDerivationRow = {|
  AccountingDerivationId: number,
  ...AccountingDerivationInsert,
|};
/**
 * We don't cache the spending counter here
 * As it would be hard to deal with rollbacks
 */
export const AccountingDerivationSchema: {
  +name: 'AccountingDerivation',
  properties: $ObjMapi<AccountingDerivationRow, ToSchemaProp>
} = {
  name: 'AccountingDerivation',
  properties: {
    AccountingDerivationId: 'AccountingDerivationId',
    KeyDerivationId: 'KeyDerivationId',
    AddressId: 'AddressId',
  }
};

export const populateCommonDb = (schemaBuilder: lf$schema$Builder) => {
  // PrivateDeriver
  schemaBuilder.createTable(PrivateDeriverSchema.name)
    .addColumn(PrivateDeriverSchema.properties.PrivateDeriverId, Type.INTEGER)
    .addColumn(PrivateDeriverSchema.properties.Bip44WrapperId, Type.INTEGER)
    .addColumn(PrivateDeriverSchema.properties.KeyDerivationId, Type.INTEGER)
    .addColumn(PrivateDeriverSchema.properties.Level, Type.INTEGER)
    .addPrimaryKey(
      ([PrivateDeriverSchema.properties.PrivateDeriverId]: Array<string>),
      true
    )
    // TODO: need to swap dependency order and move to /core/
    .addForeignKey('PrivateDeriver_Bip44Wrapper', {
      local: PrivateDeriverSchema.properties.Bip44WrapperId,
      ref: `${Bip44WrapperSchema.name}.${Bip44WrapperSchema.properties.Bip44WrapperId}`
    })
    .addForeignKey('PrivateDeriver_Bip44Derivation', {
      local: PrivateDeriverSchema.properties.KeyDerivationId,
      ref: `${KeyDerivationSchema.name}.${KeyDerivationSchema.properties.KeyDerivationId}`
    });

  // Bip44Root
  schemaBuilder.createTable(Bip44RootSchema.name)
    .addColumn(Bip44RootSchema.properties.Bip44RootId, Type.INTEGER)
    .addColumn(Bip44RootSchema.properties.KeyDerivationId, Type.INTEGER)
    .addPrimaryKey(
      ([Bip44RootSchema.properties.Bip44RootId]: Array<string>),
      true
    )
    .addForeignKey('Bip44Root_Bip44Derivation', {
      local: Bip44RootSchema.properties.KeyDerivationId,
      ref: `${KeyDerivationSchema.name}.${KeyDerivationSchema.properties.KeyDerivationId}`
    });
  // Bip44Purpose
  schemaBuilder.createTable(Bip44PurposeSchema.name)
    .addColumn(Bip44PurposeSchema.properties.Bip44PurposeId, Type.INTEGER)
    .addColumn(Bip44PurposeSchema.properties.KeyDerivationId, Type.INTEGER)
    .addPrimaryKey(
      ([Bip44PurposeSchema.properties.Bip44PurposeId]: Array<string>),
      true
    )
    .addForeignKey('Bip44Purpose_Bip44Derivation', {
      local: Bip44PurposeSchema.properties.KeyDerivationId,
      ref: `${KeyDerivationSchema.name}.${KeyDerivationSchema.properties.KeyDerivationId}`
    });
  // Bip44CoinType
  schemaBuilder.createTable(Bip44CoinTypeSchema.name)
    .addColumn(Bip44CoinTypeSchema.properties.Bip44CoinTypeId, Type.INTEGER)
    .addColumn(Bip44CoinTypeSchema.properties.KeyDerivationId, Type.INTEGER)
    .addPrimaryKey(
      ([Bip44CoinTypeSchema.properties.Bip44CoinTypeId]: Array<string>),
      true
    )
    .addForeignKey('Bip44CoinType_Bip44Derivation', {
      local: Bip44CoinTypeSchema.properties.KeyDerivationId,
      ref: `${KeyDerivationSchema.name}.${KeyDerivationSchema.properties.KeyDerivationId}`
    });
  // Bip44Account
  schemaBuilder.createTable(Bip44AccountSchema.name)
    .addColumn(Bip44AccountSchema.properties.Bip44AccountId, Type.INTEGER)
    .addColumn(Bip44AccountSchema.properties.KeyDerivationId, Type.INTEGER)
    .addPrimaryKey(
      ([Bip44AccountSchema.properties.Bip44AccountId]: Array<string>),
      true
    )
    .addForeignKey('Bip44Account_Bip44Derivation', {
      local: Bip44AccountSchema.properties.KeyDerivationId,
      ref: `${KeyDerivationSchema.name}.${KeyDerivationSchema.properties.KeyDerivationId}`
    });
  // Bip44Chain
  schemaBuilder.createTable(Bip44ChainSchema.name)
    .addColumn(Bip44ChainSchema.properties.Bip44ChainId, Type.INTEGER)
    .addColumn(Bip44ChainSchema.properties.KeyDerivationId, Type.INTEGER)
    .addColumn(Bip44ChainSchema.properties.DisplayCutoff, Type.INTEGER)
    .addPrimaryKey(
      ([Bip44ChainSchema.properties.Bip44ChainId]: Array<string>),
      true
    )
    .addForeignKey('Bip44Chain_Bip44Derivation', {
      local: Bip44ChainSchema.properties.KeyDerivationId,
      ref: `${KeyDerivationSchema.name}.${KeyDerivationSchema.properties.KeyDerivationId}`
    })
    .addNullable([
      Bip44ChainSchema.properties.DisplayCutoff,
    ]);
  // Bip44Address
  schemaBuilder.createTable(Bip44AddressSchema.name)
    .addColumn(Bip44AddressSchema.properties.Bip44AddressId, Type.INTEGER)
    .addColumn(Bip44AddressSchema.properties.KeyDerivationId, Type.INTEGER)
    .addColumn(Bip44AddressSchema.properties.AddressId, Type.INTEGER)
    .addPrimaryKey(
      ([Bip44AddressSchema.properties.Bip44AddressId]: Array<string>),
      true
    )
    .addForeignKey('Bip44Address_Bip44Derivation', {
      local: Bip44AddressSchema.properties.KeyDerivationId,
      ref: `${KeyDerivationSchema.name}.${KeyDerivationSchema.properties.KeyDerivationId}`
    })
    .addForeignKey('Bip44Address_AddressId', {
      local: Bip44AddressSchema.properties.AddressId,
      ref: `${AddressSchema.name}.${AddressSchema.properties.AddressId}`
    })
    .addIndex(
      'Bip44Address_KeyDerivation_Index',
      ([Bip44AddressSchema.properties.KeyDerivationId]: Array<string>),
      true
    );
};
