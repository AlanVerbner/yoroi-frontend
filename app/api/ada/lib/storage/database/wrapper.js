// @flow

/**
 * We have to wrap all functions that access database tables in a class
 * That way we can explicitly state which tables will be used by which functions
 * This is required because transactional queries in Lovefield
 * Need to know which tables will be accessed beforehand.
 * Getting the tables wrong will lead to either
 * - Deadlock if you lock a table which is already locked
 * - Runtime error if you access a table which is not locked
 */

import { size } from 'lodash';

export type OwnTableType = { [key: string]: string };
export type DepTableType = { [key: string]: TableClassType };
export type TableClassType = {
  ownTables: OwnTableType,
  /**
   * Recursively specify which tables will be required
   * We need to recursivley store this information
   * That way each wrapper only needs to care about the tables it specifically will access
   * and not what its dependencies will require
   */
  depTables: DepTableType,
}

/** recursively get all tables required for a database query */
export function getAllTables(tableClass: TableClassType): Set<string> {
  return new Set(_getAllTables(tableClass));
}

function _getAllTables(tableClass: TableClassType): Array<string> {
  const ownTables = Object.keys(tableClass.ownTables).map(key => tableClass.ownTables[key]);
  if (size(tableClass.depTables) === 0) {
    return ownTables;
  }
  return ownTables.concat(
    ...Object.keys(tableClass.depTables)
      .map(key => tableClass.depTables[key])
      .map(clazz => _getAllTables(clazz))
  );
}
