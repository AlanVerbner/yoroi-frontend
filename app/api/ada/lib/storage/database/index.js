// @flow

import crypto from 'crypto';
import {
  schema,
} from 'lovefield';
import type {
  lf$Database,
  lf$raw$BackStore,
} from 'lovefield';
import {
  getAllSchemaTables,
  raii,
} from './utils';
import { GetEncryptionMeta, } from './uncategorized/api/read';
import { ModifyEncryptionMeta, } from './uncategorized/api/write';

import { populateUncategorizedDb } from './uncategorized/tables';
import { populateBip44Db } from './bip44/tables';
import { populateTransactionsDb } from './transactions/tables';
import { populateWalletDb } from './wallet/tables';
import environment from '../../../../../environment';

export const loadLovefieldDB = async (
  storeType: $Values<typeof schema.DataStoreType>
): Promise<lf$Database> => {
  const db = await populateAndCreate(storeType);

  await raii(
    db,
    [
      ...getAllSchemaTables(db, GetEncryptionMeta),
      ...getAllSchemaTables(db, ModifyEncryptionMeta),
    ],
    async tx => {
      const hasMeta = await GetEncryptionMeta.exists(
        db, tx,
      );
      if (!hasMeta) {
        await ModifyEncryptionMeta.setInitial(
          db, tx,
          {
            EncryptionMetaId: 0,
            AddressSeed: environment.isJest()
              ? 1690513609
              : crypto.randomBytes(4).readUInt32BE(0),
            TransactionSeed: environment.isJest()
              ? 769388545
              : crypto.randomBytes(4).readUInt32BE(0),
            BlockSeed: environment.isJest()
              ? 371536492
              : crypto.randomBytes(4).readUInt32BE(0),
          }
        );
      }
    }
  );

  return db;
};

const populateAndCreate = async (
  storeType: $Values<typeof schema.DataStoreType>
): Promise<lf$Database> => {
  const schemaBuilder = schema.create('yoroi-schema', 2);

  populateUncategorizedDb(schemaBuilder);
  populateBip44Db(schemaBuilder);
  populateTransactionsDb(schemaBuilder);
  populateWalletDb(schemaBuilder);

  return await schemaBuilder.connect({
    storeType,
    onUpgrade,
  });
};

export async function clear(
  db: lf$Database,
): Promise<void> {
  const tx = db.createTransaction();
  await tx.begin(db.getSchema().tables());

  for (const table of db.getSchema().tables()) {
    await tx.attach(db.delete().from(table));
  }
  await tx.commit();
}

/**
 * expose dump of previous DB version so we can use it for migration
 * Note: all connection types reuse this variable unfortunately
 * since there is no way to detect the database type given just the raw back store
 */
export const dumpByVersion: { [tableName: string]: Array<any> } = {};

async function onUpgrade(
  rawDb: lf$raw$BackStore,
): Promise<void> {
  const version = rawDb.getVersion();
  if (version === 0) {
    // defaults to 0 when first time launching ever
    return;
  }
  const dump = await rawDb.dump();
  if (version === 1) {
    Object.assign(dumpByVersion, dump);
    await rawDb.dropTable('TxAddresses');
    await rawDb.dropTable('Txs');
    await rawDb.dropTable('Addresses');
  } else {
    throw new Error('unexpected version number');
  }
}
