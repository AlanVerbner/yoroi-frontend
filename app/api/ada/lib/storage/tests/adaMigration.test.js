// @flow

import '../../test-config';
import { schema, } from 'lovefield';
import { setLocalItem } from '../../../../localStorage/primitives';
import oldStorage from '../../../../../../features/yoroi_snapshots/historical-versions/1_9_0/software/localStorage';
import { RustModule } from '../../cardanoCrypto/rustLoader';
import { loadLovefieldDB } from '../database/index';
import { storagev2Migation } from '../adaMigration';
import { mockDate, filterDbSnapshot } from '../bridge/tests/common';

beforeAll(async () => {
  await RustModule.load();
});

beforeEach(() => {
  mockDate();
});

test('Migrate storage v1 to storage v2', async (done) => {
  const db = await loadLovefieldDB(schema.DataStoreType.MEMORY);
  for (const key of Object.keys(oldStorage)) {
    await setLocalItem(key, oldStorage[key]);
  }

  const migrationResult = await storagev2Migation(db);

  const keysForTest = [
    'ConceptualWallet',
    'Key',
    'Bip44Wrapper',
    'Bip44Root', // why two roots?
    'Bip44Purpose',
    'Bip44CoinType', // TODO: why is this missing?
    'Bip44Account',
    'Bip44Chain',
    'HwWalletMeta', // TODO: another test that makes use of this
  ];

  const dump = (await db.export()).tables;
  filterDbSnapshot(dump, keysForTest);
  done();
});
