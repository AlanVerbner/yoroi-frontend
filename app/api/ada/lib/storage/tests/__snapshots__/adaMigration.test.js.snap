// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Migrate storage v1 to storage v2 1`] = `
Array [
  Object {
    "Bip44Account": Array [
      Object {
        "Bip44AccountId": 1,
        "KeyDerivationId": 4,
      },
    ],
  },
  Object {
    "Bip44Chain": Array [
      Object {
        "Bip44ChainId": 1,
        "DisplayCutoff": 142,
        "KeyDerivationId": 5,
      },
      Object {
        "Bip44ChainId": 2,
        "DisplayCutoff": null,
        "KeyDerivationId": 26,
      },
    ],
  },
  Object {
    "Bip44CoinType": Array [
      Object {
        "Bip44CoinTypeId": 1,
        "KeyDerivationId": 3,
      },
    ],
  },
  Object {
    "Bip44Purpose": Array [
      Object {
        "Bip44PurposeId": 1,
        "KeyDerivationId": 2,
      },
    ],
  },
  Object {
    "Bip44Root": Array [
      Object {
        "Bip44RootId": 1,
        "KeyDerivationId": 1,
      },
    ],
  },
  Object {
    "Bip44Wrapper": Array [
      Object {
        "Bip44WrapperId": 1,
        "ConceptualWalletId": 1,
        "IsBundled": false,
        "PublicDeriverLevel": 3,
        "SignerLevel": 0,
        "Version": 2,
      },
    ],
  },
  Object {
    "ConceptualWallet": Array [
      Object {
        "CoinType": 2147485463,
        "ConceptualWalletId": 1,
        "Name": "Wallet",
      },
    ],
  },
  Object {
    "HwWalletMeta": Array [],
  },
  Object {
    "Key": Array [
      Object {
        "Hash": "a4f667dd7dd4e242fdc082540f24996a54227f6bbf0e77f735fc192fca4dbe8fcad69edddcd0ab1c5d1db62717c6b733fba5f19bcf4cda7a17770408342c4eb25373ab7e47678ab20ad428301a39324a0f789f10901ce41a128b466568f863b9b7baa989aaf09ec746615ea2400430dadec18ada015841107bd2133dc3b2af107ecea49c1b345bdb51ebda9fed6ebe6546efa20524d2dcc66c8a5605",
        "IsEncrypted": true,
        "KeyId": 1,
        "PasswordLastUpdate": "2019-07-31T13:40:16+03:00",
      },
      Object {
        "Hash": "b2938f29eb965952b03baaf010b0f5fe4a2c116a1e255274c9ace24deb72450ce1c045821c42ed11b33db3124b1551793e76b8ad187efb92b28969b92bb95964",
        "IsEncrypted": false,
        "KeyId": 2,
        "PasswordLastUpdate": null,
      },
    ],
  },
]
`;
